File: hadron-desktop/src-tauri/src/commands/export.rs
//! Report export commands

use super::common::helpers::redact_pii_basic;
use super::patterns::PatternEngineState;
use crate::export::{
    default_config_for_audience, export_report, export_report_multi, has_sensitive_content,
    sanitize_for_customer, simplify_technical_terms, ExportFormat, ReportAudience, ReportData,
    ReportSections,
};
use crate::parser::CrashFileParser;
use serde::{Deserialize, Serialize};
use tauri::State;

/// Export configuration from frontend
#[derive(Deserialize)]
pub struct ExportRequest {
    pub crash_content: String,
    pub file_name: String,
    pub format: String,
    pub audience: Option<String>,
    pub title: Option<String>,
    pub include_sections: Option<Vec<String>>,
    pub footer_text: Option<String>,
}

/// Export response with content and suggested filename
#[derive(Serialize)]
pub struct ExportResponse {
    pub content: String,
    pub suggested_filename: String,
    pub format: String,
}

/// Result of sensitive content check
#[derive(Serialize)]
pub struct SensitiveContentResult {
    pub has_sensitive: bool,
    pub warnings: Vec<String>,
    pub detected_types: Vec<String>,
}

/// Generate a report from crash content
#[tauri::command]
pub fn generate_report(
    request: ExportRequest,
    engine: State<'_, PatternEngineState>,
) -> Result<ExportResponse, String> {
    log::info!(
        "Generating {} report for: {}",
        request.format,
        request.file_name
    );

    // Parse the crash file
    let parser = CrashFileParser::new();
    let crash = parser
        .parse_content(
            &request.crash_content,
            &request.file_name,
            request.crash_content.len() as u64,
        )
        .map_err(|e| format!("Parse error: {}", e))?;

    // Get pattern match
    let engine_guard = engine
        .0
        .read()
        .map_err(|e| format!("Pattern engine lock poisoned: {}", e))?;
    let pattern_match = engine_guard.find_best_match(&crash);
    drop(engine_guard);

    // Determine audience
    let audience = match request.audience.as_deref() {
        Some("customer") => ReportAudience::Customer,
        Some("support") => ReportAudience::Support,
        Some("executive") => ReportAudience::Executive,
        _ => ReportAudience::Technical,
    };

    // Build config
    let mut config = default_config_for_audience(audience);
    config.title = request.title;

    if let Some(footer) = request.footer_text {
        config.branding.footer_text = Some(footer);
    }

    // Override sections if specified
    if let Some(sections) = request.include_sections {
        config.sections = ReportSections {
            summary: sections.contains(&"summary".to_string()),
            environment: sections.contains(&"environment".to_string()),
            exception_details: sections.contains(&"exception_details".to_string()),
            root_cause: sections.contains(&"root_cause".to_string()),
            reproduction_steps: sections.contains(&"reproduction_steps".to_string()),
            suggested_fix: sections.contains(&"suggested_fix".to_string()),
            stack_trace: sections.contains(&"stack_trace".to_string()),
            context_arguments: sections.contains(&"context_arguments".to_string()),
            database_state: sections.contains(&"database_state".to_string()),
            memory_report: sections.contains(&"memory_report".to_string()),
            system_warnings: sections.contains(&"system_warnings".to_string()),
            impact_analysis: sections.contains(&"impact_analysis".to_string()),
            test_scenarios: sections.contains(&"test_scenarios".to_string()),
            investigation_queries: sections.contains(&"investigation_queries".to_string()),
            pattern_match: sections.contains(&"pattern_match".to_string()),
        };
    }

    // Create report data
    let report_data = ReportData::from_crash(&crash, pattern_match.as_ref(), config);

    // Determine format
    let format = match request.format.to_lowercase().as_str() {
        "html" => ExportFormat::Html,
        "json" => ExportFormat::Json,
        _ => ExportFormat::Markdown,
    };

    // Generate report
    let content = export_report(&report_data, format);

    // Determine file extension
    let extension = match format {
        ExportFormat::Html => "html",
        ExportFormat::Json => "json",
        ExportFormat::Markdown => "md",
    };

    // Create suggested filename
    let base_name = std::path::Path::new(&request.file_name)
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("report");

    let suggested_filename = format!("{}_report.{}", base_name, extension);

    Ok(ExportResponse {
        content,
        suggested_filename,
        format: request.format,
    })
}

/// Get available export formats
#[tauri::command]
pub fn get_export_formats() -> Vec<serde_json::Value> {
    vec![
        serde_json::json!({
            "id": "markdown",
            "name": "Markdown",
            "extension": "md",
            "description": "Plain text with formatting, ideal for documentation"
        }),
        serde_json::json!({
            "id": "html",
            "name": "HTML",
            "extension": "html",
            "description": "Styled web page, can be opened in any browser"
        }),
        serde_json::json!({
            "id": "json",
            "name": "JSON",
            "extension": "json",
            "description": "Structured data, ideal for integrations"
        }),
    ]
}

/// Get available audience options
#[tauri::command]
pub fn get_audience_options() -> Vec<serde_json::Value> {
    vec![
        serde_json::json!({
            "id": "technical",
            "name": "Technical",
            "description": "Full details for developers"
        }),
        serde_json::json!({
            "id": "support",
            "name": "Support",
            "description": "Actionable info for support engineers"
        }),
        serde_json::json!({
            "id": "customer",
            "name": "Customer",
            "description": "Sanitized summary for end users"
        }),
        serde_json::json!({
            "id": "executive",
            "name": "Executive",
            "description": "High-level summary for management"
        }),
    ]
}

/// Preview a report without saving
#[tauri::command]
pub fn preview_report(
    crash_content: String,
    file_name: String,
    format: String,
    audience: String,
    engine: State<'_, PatternEngineState>,
) -> Result<String, String> {
    let request = ExportRequest {
        crash_content,
        file_name,
        format,
        audience: Some(audience),
        title: None,
        include_sections: None,
        footer_text: None,
    };

    let response = generate_report(request, engine)?;
    Ok(response.content)
}

/// Check content for sensitive data before sending to AI
#[tauri::command]
pub fn check_sensitive_content(content: String) -> Result<SensitiveContentResult, String> {
    log::debug!(
        "Checking content for sensitive data ({} bytes)",
        content.len()
    );

    let mut warnings = Vec::new();
    let mut detected_types = Vec::new();

    use super::common::helpers::detect_pii_types;
    let pii_types = detect_pii_types(&content);
    for pii_type in &pii_types {
        detected_types.push(pii_type.to_string());
        warnings.push(match *pii_type {
            "email" => "Email addresses detected in content".to_string(),
            "ip" => "IP addresses detected in content".to_string(),
            "token" => "API tokens or keys detected in content".to_string(),
            "path" => "User directory paths detected in content".to_string(),
            _ => format!("{} detected in content", pii_type),
        });
    }

    let has_sensitive = has_sensitive_content(&content) || !detected_types.is_empty();

    if has_sensitive && warnings.is_empty() {
        warnings.push("Potentially sensitive content detected (usernames, passwords)".to_string());
        detected_types.push("credentials".to_string());
    }

    log::info!(
        "Sensitive content check: has_sensitive={}, types={:?}",
        has_sensitive,
        detected_types
    );

    Ok(SensitiveContentResult {
        has_sensitive,
        warnings,
        detected_types,
    })
}

/// Sanitize content for a specific audience
#[tauri::command]
pub fn sanitize_content(content: String, audience: String) -> Result<String, String> {
    log::debug!("Sanitizing content for audience: {}", audience);

    let sanitized = match audience.to_lowercase().as_str() {
        "customer" | "executive" => {
            // Maximum sanitization: redact PII and simplify technical terms
            let redacted = sanitize_for_customer(&content);
            simplify_technical_terms(&redacted)
        }
        "support" => {
            // Moderate sanitization: redact PII but keep technical terms
            sanitize_for_customer(&content)
        }
        _ => {
            // Minimal sanitization (including "technical"): just redact obvious PII
            redact_pii_basic(&content).into_owned()
        }
    };

    log::info!("Content sanitized for audience: {}", audience);
    Ok(sanitized)
}

/// Multi-file report export request
#[derive(Deserialize)]
pub struct MultiExportRequest {
    pub files: Vec<ExportRequest>,
    pub format: String,
    pub combined: bool,
}

/// Generate a report from multiple crash files
#[tauri::command]
pub fn generate_report_multi(
    request: MultiExportRequest,
    engine: State<'_, PatternEngineState>,
) -> Result<ExportResponse, String> {
    log::info!(
        "Generating multi-file {} report for {} files",
        request.format,
        request.files.len()
    );

    if request.files.is_empty() {
        return Err("No files provided".to_string());
    }

    // If not combined, just process first file
    if !request.combined {
        let first = request.files.into_iter().next().unwrap();
        return generate_report(first, engine);
    }

    // Parse all crash files
    let parser = CrashFileParser::new();
    let engine_guard = engine
        .0
        .read()
        .map_err(|e| format!("Pattern engine lock poisoned: {}", e))?;

    let mut report_data_list = Vec::new();
    for file in &request.files {
        let crash = parser
            .parse_content(&file.crash_content, &file.file_name, file.crash_content.len() as u64)
            .map_err(|e| format!("Parse error for {}: {}", file.file_name, e))?;

        let pattern_match = engine_guard.find_best_match(&crash);

        let audience = match file.audience.as_deref() {
            Some("customer") => ReportAudience::Customer,
            Some("support") => ReportAudience::Support,
            Some("executive") => ReportAudience::Executive,
            _ => ReportAudience::Technical,
        };

        let config = default_config_for_audience(audience);
        let report_data = ReportData::from_crash(&crash, pattern_match.as_ref(), config);
        report_data_list.push(report_data);
    }

    drop(engine_guard);

    // Determine format
    let format = match request.format.to_lowercase().as_str() {
        "html" => ExportFormat::Html,
        "json" => ExportFormat::Json,
        _ => ExportFormat::Markdown,
    };

    // Generate combined report
    let content = export_report_multi(&report_data_list, format);

    // Determine file extension
    let extension = match format {
        ExportFormat::Html => "html",
        ExportFormat::Json => "json",
        ExportFormat::Markdown => "md",
    };

    let suggested_filename = format!("combined_report_{}.{}", request.files.len(), extension);

    Ok(ExportResponse {
        content,
        suggested_filename,
        format: request.format,
    })
}

File: hadron-electron/electron/ipc/export.ts
import { IpcMain } from 'electron'
import fs from 'fs/promises'
import path from 'path'
import crypto from 'crypto'
import * as XLSX from 'xlsx'

// ============================================================================
// Types
// ============================================================================

interface ExportRequest {
  crash_content: string
  file_name: string
  format: string
  audience?: string
  title?: string
  include_sections?: string[]
  footer_text?: string
}

interface ExportResponse {
  content: string
  suggested_filename: string
  format: string
}

interface GenericSection {
  id: string
  label: string
  content: string
}

interface GenericExportRequest {
  source_type: string
  source_name: string
  format: string
  audience?: string
  title?: string
  sections: GenericSection[]
  footer_text?: string
}

interface SensitiveContentResult {
  has_sensitive: boolean
  warnings: string[]
  detected_types: string[]
}

interface ReportMeta {
  generated_at: string
  report_id: string
  source?: string
  source_type?: string
  source_name?: string
}

// ============================================================================
// PII detection (ported from export/sanitizer.rs)
// ============================================================================

const EMAIL_RE = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g
const IP_RE = /\b(?:\d{1,3}\.){3}\d{1,3}\b/g
const TOKEN_RE = /(?:api[_-]?key|token|secret|password|pwd|auth)[^\s]*\s*[=:]\s*\S+/gi
const PATH_RE = /(?:\/home\/|\/Users\/|C:\\Users\\)[^\s/\\]+/gi
const CREDENTIAL_RE = /(?:username|user|login)\s*[=:]\s*\S+/gi

function detectPiiTypes(content: string): string[] {
  const types: string[] = []
  const tests: [RegExp, string][] = [
    [EMAIL_RE, 'email'],
    [IP_RE, 'ip'],
    [TOKEN_RE, 'token'],
    [PATH_RE, 'path'],
    [CREDENTIAL_RE, 'credentials'],
  ]
  for (const [re, label] of tests) {
    re.lastIndex = 0
    if (re.test(content)) types.push(label)
    re.lastIndex = 0
  }
  return types
}

function redactPii(content: string): string {
  TOKEN_RE.lastIndex = 0
  CREDENTIAL_RE.lastIndex = 0
  return content
    .replace(new RegExp(EMAIL_RE.source, 'g'), '[EMAIL]')
    .replace(new RegExp(IP_RE.source, 'g'), '[IP]')
    .replace(new RegExp(TOKEN_RE.source, 'gi'), (m) => m.replace(/([=:])\s*\S+/, '$1 [REDACTED]'))
    .replace(new RegExp(PATH_RE.source, 'g'), '[USER_PATH]')
    .replace(new RegExp(CREDENTIAL_RE.source, 'gi'), (m) => m.replace(/([=:])\s*\S+/, '$1 [REDACTED]'))
}

function simplifyTechnicalTerms(text: string): string {
  const replacements: [RegExp, string][] = [
    [/SubscriptOutOfBoundsError/gi, 'Application Error'],
    [/NullPointerException/gi, 'Application Error'],
    [/OutOfMemoryError/gi, 'Memory Issue'],
    [/StackOverflowError/gi, 'Application Error'],
    [/ConnectionRefusedException/gi, 'Connection Error'],
    [/SQLException/gi, 'Database Error'],
  ]
  let result = text
  for (const [re, replacement] of replacements) result = result.replace(re, replacement)
  return result
}

function applyAudienceFilter(content: string, audience: string): string {
  if (audience === 'customer' || audience === 'executive') {
    return simplifyTechnicalTerms(redactPii(content))
  }
  return redactPii(content)
}

// ============================================================================
// HTML helpers
// ============================================================================

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;')
}

const BASE_CSS = `
body{font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,sans-serif;background:#0f172a;color:#e2e8f0;margin:0;padding:2rem;line-height:1.6}
.container{max-width:900px;margin:0 auto}
h1{color:#60a5fa;border-bottom:2px solid #334155;padding-bottom:.5rem}
h2{color:#93c5fd;margin-top:1.5rem}
.meta{color:#94a3b8;font-size:.9rem;margin-bottom:1.5rem}
.section{background:#1e293b;border-radius:8px;padding:1.25rem;margin-bottom:1rem;white-space:pre-wrap;word-break:break-word}
.footer{border-top:1px solid #334155;margin-top:2rem;padding-top:1rem;color:#64748b;font-style:italic}
`

function htmlDoc(title: string, body: string): string {
  return `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1.0">
<title>${escapeHtml(title)}</title>
<style>${BASE_CSS}</style>
</head>
<body><div class="container">${body}</div></body>
</html>`
}

// ============================================================================
// Format generators
// ============================================================================

function toMarkdown(title: string, sections: GenericSection[], meta: ReportMeta, footer?: string): string {
  let md = `# ${title}\n\n`
  md += `**Generated:** ${meta.generated_at}  \n**Report ID:** ${meta.report_id}  \n`
  if (meta.source) md += `**Source:** ${meta.source}  \n`
  md += '\n'
  for (const s of sections) md += `## ${s.label}\n\n${s.content}\n\n`
  if (footer) md += `---\n*${footer}*\n`
  return md
}

function toHtml(title: string, sections: GenericSection[], meta: ReportMeta, footer?: string): string {
  const metaHtml = [
    `<div>Generated: ${escapeHtml(meta.generated_at)}</div>`,
    `<div>Report ID: ${escapeHtml(meta.report_id)}</div>`,
    meta.source ? `<div>Source: ${escapeHtml(meta.source)}</div>` : '',
  ].filter(Boolean).join('\n')

  const sectionsHtml = sections.map(s =>
    `<h2>${escapeHtml(s.label)}</h2><div class="section">${escapeHtml(s.content)}</div>`
  ).join('\n')

  const footerHtml = footer ? `<div class="footer">${escapeHtml(footer)}</div>` : ''

  return htmlDoc(title, `
<h1>${escapeHtml(title)}</h1>
<div class="meta">${metaHtml}</div>
${sectionsHtml}
${footerHtml}`)
}

function toHtmlInteractive(title: string, sections: GenericSection[], meta: ReportMeta, footer?: string): string {
  const tabButtons = sections.map((s, i) =>
    `<button class="tab${i === 0 ? ' active' : ''}" onclick="show(${i})">${escapeHtml(s.label)}</button>`
  ).join('\n')

  const tabPanels = sections.map((s, i) =>
    `<div class="panel${i === 0 ? ' active' : ''}" id="p${i}"><pre>${escapeHtml(s.content)}</pre></div>`
  ).join('\n')

  const footerHtml = footer ? `<div class="footer">${escapeHtml(footer)}</div>` : ''

  const interactiveCss = `
.tabs{display:flex;gap:.5rem;flex-wrap:wrap;margin-bottom:1rem}
.tab{background:#1e293b;border:1px solid #334155;color:#94a3b8;padding:.4rem .9rem;border-radius:6px;cursor:pointer;font-size:.9rem}
.tab.active,.tab:hover{background:#3b82f6;color:#fff;border-color:#3b82f6}
.panel{display:none;background:#1e293b;border-radius:8px;padding:1.25rem}
.panel.active{display:block}
pre{margin:0;white-space:pre-wrap;word-break:break-word;color:#e2e8f0}
`

  const script = `
function show(i){
  document.querySelectorAll('.tab').forEach((t,j)=>t.classList.toggle('active',i===j));
  document.querySelectorAll('.panel').forEach((p,j)=>p.classList.toggle('active',i===j));
}
`

  return `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1.0">
<title>${escapeHtml(title)}</title>
<style>${BASE_CSS}${interactiveCss}</style>
</head>
<body>
<div class="container">
<h1>${escapeHtml(title)}</h1>
<div class="meta">
  <div>Generated: ${escapeHtml(meta.generated_at)}</div>
  <div>Report ID: ${escapeHtml(meta.report_id)}</div>
  ${meta.source ? `<div>Source: ${escapeHtml(meta.source)}</div>` : ''}
</div>
<div class="tabs">${tabButtons}</div>
${tabPanels}
${footerHtml}
</div>
<script>${script}</script>
</body>
</html>`
}

function toJson(title: string, sections: GenericSection[], meta: ReportMeta): string {
  return JSON.stringify({ title, metadata: meta, sections }, null, 2)
}

function toTxt(title: string, sections: GenericSection[], meta: ReportMeta, footer?: string): string {
  const bar = '='.repeat(Math.min(title.length, 80))
  let txt = `${title}\n${bar}\n\nGenerated: ${meta.generated_at}\nReport ID: ${meta.report_id}\n`
  if (meta.source) txt += `Source: ${meta.source}\n`
  txt += '\n'
  for (const s of sections) {
    const dashes = '-'.repeat(Math.min(s.label.length, 80))
    txt += `${s.label}\n${dashes}\n${s.content}\n\n`
  }
  if (footer) txt += `---\n${footer}\n`
  return txt
}

function toXlsx(title: string, sections: GenericSection[], meta: ReportMeta): string {
  const wb = XLSX.utils.book_new()

  // Metadata sheet
  const metaRows = [
    ['Report Title', title],
    ['Generated', meta.generated_at],
    ['Report ID', meta.report_id],
    ...(meta.source ? [['Source', meta.source]] : []),
  ]
  XLSX.utils.book_append_sheet(wb, XLSX.utils.aoa_to_sheet(metaRows), 'Info')

  // One sheet per section (max 31-char sheet names, Excel limit)
  for (const section of sections) {
    const sheetName = section.label.slice(0, 31).replace(/[[\]\\/?*:]/g, '_')
    const rows = section.content.split('\n').map(line => [line])
    XLSX.utils.book_append_sheet(wb, XLSX.utils.aoa_to_sheet(rows), sheetName)
  }

  const buf = XLSX.write(wb, { type: 'buffer', bookType: 'xlsx' })
  return Buffer.from(buf).toString('base64')
}

// ============================================================================
// Report dispatch
// ============================================================================

function makeReportId(): string {
  return crypto.randomUUID()
}

function formatReport(
  title: string,
  sections: GenericSection[],
  format: string,
  meta: ReportMeta,
  footer?: string
): { content: string; extension: string } {
  switch (format.toLowerCase()) {
    case 'html':
      return { content: toHtml(title, sections, meta, footer), extension: 'html' }
    case 'html_interactive':
      return { content: toHtmlInteractive(title, sections, meta, footer), extension: 'html' }
    case 'json':
      return { content: toJson(title, sections, meta), extension: 'json' }
    case 'txt':
    case 'text':
      return { content: toTxt(title, sections, meta, footer), extension: 'txt' }
    case 'xlsx':
    case 'excel':
      return { content: toXlsx(title, sections, meta), extension: 'xlsx' }
    default:
      return { content: toMarkdown(title, sections, meta, footer), extension: 'md' }
  }
}

// ============================================================================
// IPC handler registration
// ============================================================================

export function registerExportHandlers(ipc: IpcMain): void {
  // File write aliases used by ExportDialog / ExportMenu
  ipc.handle('write_export_text', async (_e, args: { path: string; content: string }) => {
    await fs.writeFile(args.path, args.content, 'utf-8')
  })

  ipc.handle('write_export_bytes', async (_e, args: { path: string; data: number[] }) => {
    await fs.writeFile(args.path, Buffer.from(args.data))
  })

  // Static data
  ipc.handle('get_export_formats', () => [
    { id: 'markdown', name: 'Markdown', extension: 'md', description: 'Plain text with formatting, ideal for documentation' },
    { id: 'html', name: 'HTML', extension: 'html', description: 'Styled web page, can be opened in any browser' },
    { id: 'html_interactive', name: 'Interactive HTML', extension: 'html', description: 'Interactive page with tabbed navigation per section' },
    { id: 'json', name: 'JSON', extension: 'json', description: 'Structured data, ideal for integrations' },
    { id: 'txt', name: 'Plain Text', extension: 'txt', description: 'Simple text format without formatting' },
    { id: 'xlsx', name: 'Excel (XLSX)', extension: 'xlsx', description: 'Multi-sheet spreadsheet with tabbed sections', isBinary: true },
  ])

  ipc.handle('get_audience_options', () => [
    { id: 'technical', name: 'Technical', description: 'Full details for developers' },
    { id: 'support', name: 'Support', description: 'Actionable info for support engineers' },
    { id: 'customer', name: 'Customer', description: 'Sanitized summary for end users' },
    { id: 'executive', name: 'Executive', description: 'High-level summary for management' },
  ])

  // PII detection and sanitization
  ipc.handle('check_sensitive_content', (_e, args: { content: string }): SensitiveContentResult => {
    const types = detectPiiTypes(args.content)
    const WARNINGS: Record<string, string> = {
      email: 'Email addresses detected in content',
      ip: 'IP addresses detected in content',
      token: 'API tokens or keys detected in content',
      path: 'User directory paths detected in content',
      credentials: 'Credentials detected in content',
    }
    return {
      has_sensitive: types.length > 0,
      warnings: types.map(t => WARNINGS[t] ?? `${t} detected in content`),
      detected_types: types,
    }
  })

  ipc.handle('sanitize_content', (_e, args: { content: string; audience: string }): string => {
    return applyAudienceFilter(args.content, args.audience)
  })

  // generate_report — for crash sourceType in ExportDialog
  ipc.handle('generate_report', (_e, args: { request: ExportRequest }): ExportResponse => {
    const { request } = args
    const audience = request.audience ?? 'technical'
    const now = new Date().toISOString().replace('T', ' ').slice(0, 19)
    const title = request.title ?? 'Crash Analysis Report'
    const baseName = path.basename(request.file_name, path.extname(request.file_name))

    const body = applyAudienceFilter(request.crash_content, audience)
    const sections: GenericSection[] = [{ id: 'analysis', label: 'Analysis', content: body }]
    const meta: ReportMeta = {
      generated_at: now,
      report_id: makeReportId(),
      source: request.file_name,
    }

    const { content, extension } = formatReport(title, sections, request.format, meta, request.footer_text)
    return { content, suggested_filename: `${baseName}_report.${extension}`, format: request.format }
  })

  // export_generic_report — for code/sentry/jira analyzers
  ipc.handle('export_generic_report', (_e, args: { request: GenericExportRequest }): ExportResponse => {
    const { request } = args
    const now = new Date().toISOString().replace('T', ' ').slice(0, 19)
    const title = request.title ?? `${request.source_type} Report`
    const baseName = request.source_name.replace(/[/\\:*?"<>|]/g, '_')

    const audience = request.audience ?? 'technical'
    const sections: GenericSection[] = request.sections.map(s => ({
      ...s,
      content: applyAudienceFilter(s.content, audience),
    }))

    const meta: ReportMeta = {
      generated_at: now,
      report_id: makeReportId(),
      source: `${request.source_type} (${request.source_name})`,
      source_type: request.source_type,
      source_name: request.source_name,
    }

    const { content, extension } = formatReport(title, sections, request.format, meta, request.footer_text)
    return {
      content,
      suggested_filename: `${baseName}_${request.source_type}_report.${extension}`,
      format: request.format,
    }
  })

  // preview_report — returns HTML string for in-app preview
  ipc.handle('preview_report', (_e, args: {
    crash_content: string
    file_name: string
    format: string
    audience: string
    title?: string
  }): string => {
    const body = applyAudienceFilter(args.crash_content, args.audience ?? 'technical')
    const sections: GenericSection[] = [{ id: 'analysis', label: 'Analysis', content: body }]
    const meta: ReportMeta = {
      generated_at: new Date().toISOString().replace('T', ' ').slice(0, 19),
      report_id: makeReportId(),
      source: args.file_name,
    }
    return toHtml(args.title ?? 'Report Preview', sections, meta)
  })
}
